src/security/PeerOptions.h

```diff
@@ -145,15 +150,88 @@ class PeerOptions
     bool encryptTransport = false;
 };
 
+// TODO: Move this declaration.
+/// A combination of PeerOptions and the corresponding Context. Used by Squid
+/// TLS client code.
+class PeerContext: public RefCountable
+{
+public:
+    explicit PeerContext(ConfigParser &);
+
+    /// creates TLS context reflecting the configured options
+    void open();
+
+    PeerOptions options; ///< context configuration
+    ContextPointer raw; ///< context configured using options
+
+    /// restrict usage to matching transactions
+    std::unique_ptr<ACLList> preconditions;
+};
+
+// TODO: Move this declaration.
+/// Manages PeerContext objects representing all
+/// tls_outgoing_options_for_retries directives in squid.conf.
+class PeerContexts
+{
+public:
+    /// parses a single tls_outgoing_options_for_retries directive
+    void parseOneDirective(ConfigParser &);
+
+    /// opens all configured contexts; \sa PeerContext::open()
+    void open();
+
+    /// transaction-matching PeerContext (or nil)
+    PeerContextPointer findContext(ACLChecklist &) const;
+
+    /// configured contexts in squid.conf directives order
+    std::vector< PeerContextPointer, PoolingAllocator<PeerContextPointer> > contexts;
+};
+
+// XXX: Remove this shim after upgrading legacy code to store PeerContext
+// objects instead of disjoint PeerOptons and Context objects.
+/// A combination of PeerOptions and the corresponding Context. Used by Squid
+/// TLS client code.
+class FuturePeerContext: public RefCountable
+{
+public:
+    FuturePeerContext(PeerOptions &o, const ContextPointer &c): options(o), raw(c)
+    {
+    }
+
+    PeerOptions &options; ///< TLS context configuration
+    const ContextPointer &raw; ///< TLS context configured using options
```

Comment:
Q1: Should I post a PR that fixes bug 5293 while using FuturePeerContext shim as necessary to prevent lots of code changes that have pretty much nothing to do with that bug? The alternative is to post that bug-fixing PR without the shim (and with a lot of noise).

This shim adds a small performance overhead (we need to allocate those trivial shim objects whenever a new TLS connection is created), but if we promptly remove it in a followup PR, then it is not a big problem.

Reply:
I trust this to be done. Thanks!

Reply:
> OK, we will try to do a small fix using a shim to facilitate v6 backporting. That matches the first change described in the https://github.com/squid-cache/squid/pull/1456#pullrequestreview-1575490152 for this Draft PR.

Posted as #1895. Once that PR lands in master/v7, I will update this PR accordingly, of course.